README.md
<p align="center">
  <img width="480" src="assets/logo.png"/>
</p>

[![CocoaPods](https://cocoapod-badges.herokuapp.com/v/SpaceBunny/badge.svg)](http://www.cocoapods.org/?q=spacebunny)
[![Build Status](https://travis-ci.org/andreamazz/AMScrollingNavbar.svg)](https://travis-ci.org/space-bunny/swift-sdk)
[![codecov.io](https://codecov.io/github/space-bunny/swift-sdk/coverage.svg?branch=master)](https://codecov.io/github/space-bunny/swift-sdk?branch=master)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
![Swift 2.3](https://img.shields.io/badge/swift-2.3-orange.svg)

[SpaceBunny](http://spacebunny.io) is the IoT platform that makes it easy for you and your devices to send and exchange messages with a server or even with each other. You can store the data, receive timely event notifications, monitor live streams and remotely control your devices. Easy to use, and ready to scale at any time.

##Installation

###CocoaPods
```
use_frameworks!

pod 'SpaceBunny'
```

##Usage

###Connection
Configure the instance of the SpaceBunny's `Client` with a valid Device Key:
```swift
let client = SpaceBunnyClient(deviceKey: "some-device-key")
```
Set up the client if needed:
```swift
client.useSSL = true
client.delegate = self
```
Connect the client to the platform. You can provide a completion block called when the connection is either successfull or failed (in this case you are provided with the info about the error encountered)
```swift
client.connect() { error in
  if let error == error {
    print(error)
  }
}
```

###Publishing
You can publish a message by providing the channel name. You can retrieve the channels associated with the device via the `channels` property of a client:
```swift
do {
  try client.publishOn("data", message: "test")
} catch {
  print("Unable to publish")
}
```

###Subscribing
You can subscribe to the current device's inbox and receive messages as soon as they are avaialble by either providing a closure to the `subscribe` function or by implementing the `client(_, didReceiveMessage: topic:)` 
```swift
do {
  try client.subscribe() { (message, topic) in
    print("Received message: \(message ?? "No message")")
  }
} catch {
  print("Unable to subscribe")
}
```

###Delegate methods
```swift
// Called when a connection to the platform is established
func spaceBunnyClient(client: SpaceBunnyClient, didConnectTo host: String, port: Int)
   
// Called when a connection to the platform is ended or failed
func spaceBunnyClient(client: SpaceBunnyClient, didDisconnectWithError error: NSError?)
   
// Called when a new message is received on the device's inbox
func spaceBunnyClient(client: SpaceBunnyClient, didReceiveMessage message: String?, topic: String)
   
// Called when a new message is sent
func spaceBunnyClient(client: SpaceBunnyClient, didPublishMessage message: String?, topic: String)
   
// Called when a the client subscribes to the device's inbox
func spaceBunnyClientiDidSubscribe(client: SpaceBunnyClient)

// Called when a the client unsubscribes from the device's inbox
func spaceBunnyClientDidUnsubscribe(client: SpaceBunnyClient)
```

## MIT License
    The MIT License (MIT)

    Copyright (c) 2016 Fancy Pixel S.r.l.

    Permission is hereby granted, free of charge, to any person obtaining a copy of
    this software and associated documentation files (the "Software"), to deal in
    the Software without restriction, including without limitation the rights to
    use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
    the Software, and to permit persons to whom the Software is furnished to do so,
    subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
    FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
    COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
    IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
    CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
`SpaceBunny` adheres to [Semantic Versioning](http://semver.org/).

- `0.1.x` Releases - [0.1.0](#010) | [0.1.1](#011) | [0.1.2](#012)  

---

## [0.1.2](https://github.com/space-bunny/swift-sdk/releases/tag/0.1.2)

Updated CocoaMQTT dependency  

## [0.1.1](https://github.com/space-bunny/swift-sdk/releases/tag/0.1.1)

###Added

- Error handling when the device key is not valid

## [0.1.0](https://github.com/space-bunny/swift-sdk/releases/tag/0.1.0)

Initial release


Source/Client.swift
import Foundation
import ObjectMapper
import CocoaMQTT

let EndpointScheme = "https"
let EndpointUrl = "api.spacebunny.io"

/// Exception raised by SpaceBunny's `Client`
enum SpaceBunnyError: ErrorType {
  /// The client was unexpectedly not connected
  case NotConnected
}

/** Connection state. Possible values:
 - `Connecting`
 - `Connected`
 - `Disconnected`
 */
@objc public enum SpaceBunnyState: Int {
  /// The client is set up and connecting to the platform
  case Connecting
  /// The client currently connected
  case Connected
  /// The client disconnected
  case Disconnected
}

/// SpaceBunny delegate protocol.
@objc public protocol SpaceBunnyDelegate: NSObjectProtocol {

  /**
 Called when a connection to the platform is established
   
   - parameter client: the SpaceBunnyClient instance
   - parameter host: the connection host
   - parameter host: the connection port
 */
  optional func spaceBunnyClient(client: SpaceBunnyClient, didConnectTo host: String, port: Int)

  /**
   Called when a connection to the platform is ended or failed

   - parameter client: the SpaceBunnyClient instance
   - parameter error: the error
   */
  optional func spaceBunnyClient(client: SpaceBunnyClient, didDisconnectWithError error: NSError?)

  /**
   Called when a new message is received on the device's inbox

   - parameter client: the SpaceBunnyClient instance
   - parameter message: the message
   - parameter topic: the topic
   */
  optional func spaceBunnyClient(client: SpaceBunnyClient, didReceiveMessage message: String?, topic: String)

  /**
   Called when a new message is sent

   - parameter client: the `SpaceBunnyClient` instance
   - parameter message: the message
   - parameter topic: the topic
   */
  optional func spaceBunnyClient(client: SpaceBunnyClient, didPublishMessage message: String?, topic: String)

  /**
   Called when a the client subscribes to the device's inbox

   - parameter client: the SpaceBunnyClient instance
   */
  optional func spaceBunnyClientDidSubscribe(client: SpaceBunnyClient)

  /**
   Called when a the client unsubscribes from the device's inbox

   - parameter client: the SpaceBunnyClient instance
   */
  optional func spaceBunnyClientDidUnsubscribe(client: SpaceBunnyClient)
}

/**
 SpaceBunny Client for device communications. It represents a single device and is used to
 send data on the device channels and subscribe to the device's inbox.
 */
@objc public class SpaceBunnyClient: NSObject {

  private var endpointScheme = ""
  private var endpointURLString = ""
  private var endpointPort: Int?
  internal var mqttClient: CocoaMQTT?
  private var onReceive: ((String?, String) -> Void)? = nil
  private var onConnect: (NSError? -> Void)? = nil

  /// The device key. It must be set in the `SpaceBunnyClient` initializer
  public private(set) var deviceKey: String?

  /// And array of `Channels` associated with the device. The information is retrieved upon connection
  public private(set) var channels: [Channel]? = nil

  /// The configuration parameters for the device, retrieved upon connection
  public private(set) var configuration: Configuration?

  /// Enables the secure connection to the platform
  public var useSSL = false

  /// The SpaceBunnyClient delegate
  public var delegate: SpaceBunnyDelegate?

  /**
   Convenience initializer. A valid device key must be provided. Get a valid device key on your 
   SpaceBunny's dashboard (http://spacebunny.io).
   An optional endpoint can be specified, if it differs from the default one (https://api.spacebunny.io)
   
   - parameter deviceKey: The device key
   - parameter endpointScheme: Optional endpoint URL scheme
   - parameter endpointUrl: Optional endpoint URL
   - parameter endpointPort: Optional endpoint URL port
   */
  public init(deviceKey: String, endpointScheme: String, endpointUrl: String, endpointPort: NSNumber?) {
    self.endpointURLString = endpointUrl
    self.endpointPort = endpointPort?.integerValue
    self.endpointScheme = endpointScheme
    self.deviceKey = deviceKey
  }

  /**
   Convenience initializer. A valid device key must be provided. Get a valid device key on your
   SpaceBunny's dashboard (http://spacebunny.io).

   - parameter deviceKey: The device key
   */
  public init(deviceKey: String) {
    self.endpointURLString = EndpointUrl
    self.endpointScheme = EndpointScheme
    self.deviceKey = deviceKey
  }

  /**
   Connect to the platform. The connection process retrieves the device configurations, the channels list and
   establishes a MQTT connection.

   - parameter completion: A block called when the connection is established. 
   If the connection failed, a non nil error is provided
   */
  public func connect(completion: ((NSError?) -> Void)? = nil) {
    guard let url = endpointURL() else {
      completion?(NSError(domain: "SpaceBunnyClient", code: 400, userInfo: ["error": "Non valid URL"]));
      return
    }

    let request = NSMutableURLRequest(URL: url)
    request.setValue(deviceKey, forHTTPHeaderField: "Device-Key")
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")

    NSURLSession.sharedSession().dataTaskWithRequest(request) { (data, response, error) in
      if let error = error {
        completion?(error)
      }
      if let httpResponse = response as? NSHTTPURLResponse where httpResponse.statusCode == 401 {
        completion?(NSError(domain: "SpaceBunnyClient", code: 401, userInfo: ["error": "Unauthorized. Please check your Device Key"]));
        return
      }
      if let data = data {
        do {
          let json = try NSJSONSerialization.JSONObjectWithData(data, options: .AllowFragments)

          if let json = json as? NSDictionary, connection = json["connection"] as? NSDictionary {
            self.configuration = Mapper<Configuration>().map(connection)
          }
          if let json = json as? NSDictionary, channels = json["channels"] as? Array<NSDictionary> {
            self.channels = channels.flatMap { Mapper<Channel>().map($0) }
          }
          self.mqttConnect(completion)
        } catch let error as NSError {
          completion?(error)
        }
      }
      }.resume()
  }

  /**
   Publish a message on a given channel.

   - parameter channel: The name of the channel
   - parameter message: The message string
   - Throws `NotConnected` exception if no connection is available.
   */
  public func publishOn(channel: String, message: String) throws {
    guard let configuration = configuration, username = configuration.username where mqttClient?.connState == .CONNECTED else {
      throw SpaceBunnyError.NotConnected
    }

    mqttClient?.publish("\(username)/\(channel)", withString: message)
  }

  /**
   Subscribe to the device inbox. It throws an exception if no connection is available.

   - Throws `NotConnected` exception if no connection is available.
   - parameter onReceive: An optional block called when a new message is available.
   - seealso `CocoaMQTTMessage` for the message object returned
   */
  public func subscribe(onReceive: ((message: String?, topic: String) -> Void)?) throws {
    guard let configuration = configuration, username = configuration.username where mqttClient?.connState == .CONNECTED else {
      throw SpaceBunnyError.NotConnected
    }

    self.onReceive = onReceive
    mqttClient?.subscribe("\(username)/inbox")
  }

  /**
   Unsubscribe from the device inbox. It throws an exception if no connection is available.

   - Throws `NotConnected` exception if no connection is available.
   */
  public func unsubscribe() throws {
    guard let configuration = configuration, username = configuration.username where mqttClient?.connState == .CONNECTED else {
      throw SpaceBunnyError.NotConnected
    }

    onReceive = nil
    mqttClient?.unsubscribe("\(username)/inbox")
  }

  /**
   Connection status. Returns a value of type `SpaceBunnyState`

   - returns: SpaceBunnyState
   - seealso:
   SpaceBunnyState
   */
  public func status() -> SpaceBunnyState {
    guard let mqttClient = mqttClient else { return .Disconnected }

    switch mqttClient.connState {
    case CocoaMQTTConnState.INIT:
      return .Disconnected
    case CocoaMQTTConnState.CONNECTING:
      return .Connecting
    case CocoaMQTTConnState.CONNECTED:
      return .Connected
    case CocoaMQTTConnState.DISCONNECTED:
      return .Disconnected
    }
  }

  /**
   Disconnect from the platform
   */
  public func disconnect() {
    mqttClient?.disconnect()
  }

  func endpointURL() -> NSURL? {
    var port = ""
    if let endpointPort = endpointPort {
      port = ":\(endpointPort)"
    }
    return NSURL(string: "\(endpointScheme)://\(endpointURLString)\(port)/v1/device_configurations")
  }

  func mqttConnect(completion: (NSError? -> Void)? = nil) {
    guard let configuration = configuration, username = configuration.username, password = configuration.password, host = configuration.host else { return }
    mqttClient = CocoaMQTT(clientId: username, host: host, port: UInt16((useSSL ? configuration.sslport : configuration.port) ?? 1883))

    if let mqtt = mqttClient {
      onConnect = completion
      mqtt.secureMQTT = useSSL
      mqtt.username = "\(configuration.vhost ?? ""):\(username)"
      mqtt.password = password
      mqtt.keepAlive = 90
      mqtt.delegate = self
      mqtt.connect()
    }
  }

}

extension SpaceBunnyClient: CocoaMQTTDelegate {
  public func mqtt(mqtt: CocoaMQTT, didConnect host: String, port: Int) {
    onConnect?(nil)
    delegate?.spaceBunnyClient?(self, didConnectTo: host, port: port)
  }

  public func mqtt(mqtt: CocoaMQTT, didPublishMessage message: CocoaMQTTMessage, id: UInt16) {
    delegate?.spaceBunnyClient?(self, didPublishMessage: message.string, topic: message.topic)
  }

  public func mqtt(mqtt: CocoaMQTT, didReceiveMessage message: CocoaMQTTMessage, id: UInt16) {
    onReceive?(message.string, message.topic)
    delegate?.spaceBunnyClient?(self, didReceiveMessage: message.string, topic: message.topic)
  }

  public func mqtt(mqtt: CocoaMQTT, didSubscribeTopic topic: String) {
    delegate?.spaceBunnyClientDidSubscribe?(self)
  }

  public func mqtt(mqtt: CocoaMQTT, didUnsubscribeTopic topic: String) {
    delegate?.spaceBunnyClientDidUnsubscribe?(self)
  }

  public func mqttDidDisconnect(mqtt: CocoaMQTT, withError err: NSError?) {
    delegate?.spaceBunnyClient?(self, didDisconnectWithError: err)
  }

  public func mqtt(mqtt: CocoaMQTT, didConnectAck ack: CocoaMQTTConnAck) {}
  public func mqttDidPing(mqtt: CocoaMQTT) {}
  public func mqttDidReceivePong(mqtt: CocoaMQTT) {}
  public func mqtt(mqtt: CocoaMQTT, didPublishAck id: UInt16) {}

}


Source/Configuration.swift
import Foundation
import ObjectMapper

/**
 SpaceBunny Configuration object. It holds the info of a single device and is retrieved during the
 client connection
 */
public class Configuration: Mappable {

  /// Host for the MQTT connection
  public var host: String?
  /// Username for the MQTT connection
  public var username: String?
  /// Password for the MQTT connection
  public var password: String?
  /// The device name
  public var name: String?
  /// Port for the MQTT connection
  public var port: UInt?
  /// Port for the secure MQTT connection
  public var sslport: UInt?
  /// vHost for the MQTT connection
  public var vhost: String?

  /**
   Manually create a new Configuration

   - parameter host: The connection host
   - parameter username: The connection username
   - parameter password: The connection password
   - parameter name: The device name
   - parameter port: The connection port
   - parameter sslport: The secure connection port
   - parameter vhost: The connection vHost
   */
  public required init(host: String, username: String, password: String, name: String, port: UInt, sslport: UInt, vhost: String) {
    self.host = host
    self.username = username
    self.password = password
    self.name = name
    self.port = port
    self.sslport = sslport
    self.vhost = vhost
  }

  /**
   ObjectMapper constructor
   
   - parameter map: The Map object
   */
  public required init?(_ map: Map) {
    mapping(map)
  }

  // MARK: - API Mapping

  /**
   ObjectMapper mapping

   - parameter map: The Map object
   */
  public func mapping(map: Map) {
    host <- map["host"]
    username <- map["device_id"]
    password <- map["secret"]
    name <- map["device_name"]
    port <- map["protocols.mqtt.port"]
    sslport <- map["protocols.mqtt.ssl_port"]
    vhost <- map["vhost"]
  }
}


Demo/SpaceBunny/AppDelegate.swift
//
//  AppDelegate.swift
//  SpaceBunny
//
//  Created by Andrea Mazzini on 08/05/16.
//  Copyright © 2016 Andrea Mazzini. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(application: UIApplication, didFinishLaunchingWithOptions launchOptions: [NSObject: AnyObject]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



Demo/SpaceBunny/ViewController.swift
//
//  ViewController.swift
//  SpaceBunny
//
//  Created by Andrea Mazzini on 08/05/16.
//  Copyright © 2016 Andrea Mazzini. All rights reserved.
//

import UIKit
import SpaceBunny

class ViewController: UIViewController {

  // Create a client with a given device key. Get your key from your SpaceBunny's dashboard
  let client = SpaceBunnyClient(deviceKey: "1377cd37-dae6-4a44-b1b9-478c32df2f23:xT6JysSZsMWyVM5EtU8_nw")

  override func viewDidLoad() {
    super.viewDidLoad()
  }

  func connectAndSubscribe() {
    // Set the SpaceBunnyDelegate
    client.delegate = self

    // Enable SSL
    client.useSSL = true

    // Connect to the platform.
    client.connect() { [weak self] error in
      // Completion handler provides an optional error.
      guard error == .None else { print(error); return }
      self?.subscribe()
    }
  }

  func subscribe() {
    // Subscribe to the device's inbox. Can throw
    do {
      try client.subscribe() { (message, topic) in
        print("Received message: \(message ?? "No message")")
      }
    } catch {
      print("no connection")
    }
  }

  @IBAction func connectAction() {
    connectAndSubscribe()
  }

  @IBAction func buttonAction() {
    // Publish a message on a given topic. Can throw
    do {
      try client.publishOn("data", message: "test")
    } catch {
      print("no connection")
    }
  }

}

// MARK: - SpaceBunnyDelegate

extension ViewController: SpaceBunnyDelegate {

  func spaceBunnyClient(client: SpaceBunnyClient, didConnectTo host: String, port: Int) {
    print("Connection successful")
  }

  func spaceBunnyClient(client: SpaceBunnyClient, didDisconnectWithError error: NSError?) {
    print("Disconnected")
    if let error = error {
      print("With error: \(error)")
    }
  }

}



Demo/SpaceBunnyTests/SpaceBunnyTests.swift
//
//  SpaceBunnyTests.swift
//  SpaceBunnyTests
//
//  Created by Andrea Mazzini on 08/05/16.
//  Copyright © 2016 Andrea Mazzini. All rights reserved.
//

import XCTest
@testable import SpaceBunny
import CocoaMQTT
import Mockingjay

class MockMQTTClient: CocoaMQTT {

  var expectedState = CocoaMQTTConnState.CONNECTED
  internal var didPublish = false
  internal var isSubscribed = false

  init(expectedState state: CocoaMQTTConnState) {
    expectedState = state
    super.init(clientId: "")
  }

  override func subscribe(topic: String, qos: CocoaMQTTQOS) -> UInt16 {
    isSubscribed = true
    return 0
  }

  override func unsubscribe(topic: String) -> UInt16 {
    isSubscribed = false
    return 0
  }

  override func publish(topic: String, withString string: String, qos: CocoaMQTTQOS, retained: Bool, dup: Bool) -> UInt16 {
    didPublish = true
    return 0
  }

  override var connState: CocoaMQTTConnState {
    get {
      return expectedState
    }
    set { }
  }

}

class MockClient: SpaceBunnyClient {
  var expectedState = CocoaMQTTConnState.CONNECTED

  convenience init(deviceKey: String, expectedState state: CocoaMQTTConnState) {
    self.init(deviceKey: "")
    expectedState = state
  }

  override func mqttConnect(completion: (NSError? -> Void)?) {
    self.mqttClient = MockMQTTClient(expectedState: expectedState)
    completion?(nil)
  }
}

class SpaceBunnyTests: XCTestCase {

  func successMock(url: String = "https://api.spacebunny.io/v1/device_configurations") {
    let protocols = ["mqtt": ["port": 1883, "ssl_port": 8883]]
    let channels = [[ "id": "12345", "name": "data" ], [ "id": "54321", "name": "alarms" ]]
    let config = ["host": "mock.host", "device_id": "device", "device_name": "Some device", "secret": "s3cr3t", "vhost": "mock.vhost", "protocols": protocols]
    let body = [ "connection": config, "channels": channels ]
    stub(http(.GET, uri: url), builder: json(body))
  }

  func failueMock() {
    let error = NSError(domain: "MockingjayTests", code: 401, userInfo: nil)
    stub(http(.GET, uri: "https://api.spacebunny.io/v1/device_configurations"), builder: failure(error))
  }

  func malformedResponseMock() {
    stub(http(.GET, uri: "https://api.spacebunny.io/v1/device_configurations"), builder: http(200, headers: nil, data: NSData()))
  }

  override func setUp() {
    super.setUp()
  }

  override func tearDown() {
    super.tearDown()
  }

  func testConfiguration() {
    successMock()
    let expectation = self.expectationWithDescription("fetch config")

    let client = MockClient(deviceKey: "some-key")
    client.connect() { _ in
      XCTAssertTrue(client.configuration?.host == "mock.host")
      XCTAssertTrue(client.configuration?.vhost == "mock.vhost")
      XCTAssertTrue(client.configuration?.username == "device")
      XCTAssertTrue(client.configuration?.password == "s3cr3t")
      XCTAssertTrue(client.configuration?.name == "Some device")
      XCTAssertTrue(client.configuration?.port == 1883)
      XCTAssertTrue(client.configuration?.sslport == 8883)
      XCTAssertTrue(client.channels?.count == 2)
      let channel = client.channels?.first
      XCTAssertTrue(channel?.name == "data")
      XCTAssertTrue(channel?.id == "12345")
      expectation.fulfill()
    }
    self.waitForExpectationsWithTimeout(0.5, handler: nil)
  }

  func testUnauthorized() {
    failueMock()
    let expectation = self.expectationWithDescription("fetch config")

    let client = MockClient(deviceKey: "some-wrong-key")
    client.connect() { error in
      XCTAssertTrue(error?.code == 401)
      expectation.fulfill()
    }
    self.waitForExpectationsWithTimeout(0.5, handler: nil)
  }

  func testCustomEndpoint() {
    successMock("https://endpoint.com:8080/v1/device_configurations")
    let expectation = self.expectationWithDescription("fetch config")

    let client = MockClient(deviceKey: "some-key", endpointScheme: "https", endpointUrl: "endpoint.com", endpointPort: 8080)
    client.connect() { _ in
      XCTAssertTrue(client.configuration?.host == "mock.host")
      expectation.fulfill()
    }
    self.waitForExpectationsWithTimeout(0.5, handler: nil)
  }

  func testInvalidEndpoint() {
    let expectation = self.expectationWithDescription("fetch config")

    let client = MockClient(deviceKey: "some-key", endpointScheme: "^^^", endpointUrl: "???", endpointPort: nil)
    client.connect() { error in
      XCTAssertNotNil(error)
      expectation.fulfill()
    }
    self.waitForExpectationsWithTimeout(0.5, handler: nil)
  }

  func testMalformedResponse() {
    let expectation = self.expectationWithDescription("fetch config")

    malformedResponseMock()
    let client = MockClient(deviceKey: "some-key")
    client.connect() { error in
      XCTAssertNotNil(error)
      expectation.fulfill()
    }
    self.waitForExpectationsWithTimeout(0.5, handler: nil)
  }

  func testAuthenticationFailure() {
    failueMock()
    let expectation = self.expectationWithDescription("auth error")

    let client = MockClient(deviceKey: "some-wrong-key")
    client.connect() { error in
      XCTAssertNotNil(error)
      expectation.fulfill()
    }
    self.waitForExpectationsWithTimeout(0.5, handler: nil)
  }

  func testPublishOnSuccess() {
    let expectation = self.expectationWithDescription("publish success")

    let client = MockClient(deviceKey: "some-key", expectedState: .CONNECTED)
    successMock()
    client.connect() { _ in
      try! client.publishOn("data", message: "Hello")
      if let mqtt = client.mqttClient as? MockMQTTClient {
        XCTAssertTrue(mqtt.didPublish)
      }
      expectation.fulfill()
    }
    self.waitForExpectationsWithTimeout(0.5, handler: nil)
  }

  func testPublishFailure() {
    let client = MockClient(deviceKey: "some-key", expectedState: .DISCONNECTED)
    successMock()
    XCTAssertThrowsError(try client.publishOn("data", message: "Hello"))
  }

  func testSubscribeSuccess() {
    let expectation = self.expectationWithDescription("subscribe success")

    let client = MockClient(deviceKey: "some-key", expectedState: .CONNECTED)
    successMock()
    client.connect() { _ in
      try! client.subscribe(nil)
      if let mqtt = client.mqttClient as? MockMQTTClient {
        XCTAssertTrue(mqtt.isSubscribed)
      }
      expectation.fulfill()
    }
    self.waitForExpectationsWithTimeout(0.5, handler: nil)
  }

  func testSubscribeFailure() {
    let client = MockClient(deviceKey: "some-key", expectedState: .DISCONNECTED)
    successMock()
    XCTAssertThrowsError(try client.subscribe(nil))
  }

  func testUnsubscribeSuccess() {
    let expectation = self.expectationWithDescription("subscribe success")

    let client = MockClient(deviceKey: "some-key", expectedState: .CONNECTED)
    successMock()
    client.connect() { _ in
      try! client.subscribe(nil)
      try! client.unsubscribe()
      if let mqtt = client.mqttClient as? MockMQTTClient {
        XCTAssertFalse(mqtt.isSubscribed)
      }
      expectation.fulfill()
    }
    self.waitForExpectationsWithTimeout(0.5, handler: nil)
  }

  func testUnsubscribeFailure() {
    let client = MockClient(deviceKey: "some-key", expectedState: .DISCONNECTED)
    successMock()
    XCTAssertThrowsError(try client.unsubscribe())
  }

}

class ConfigurationTests: XCTestCase {

  func testCreateManually() {
    let subject = Configuration(host: "host", username: "username", password: "password", name: "name", port: 1883, sslport: 8883, vhost: "vhost")
    XCTAssertTrue(subject.host == "host")
    XCTAssertTrue(subject.vhost == "vhost")
    XCTAssertTrue(subject.username == "username")
    XCTAssertTrue(subject.password == "password")
    XCTAssertTrue(subject.name == "name")
    XCTAssertTrue(subject.port == 1883)
    XCTAssertTrue(subject.sslport == 8883)
  }

}